massa-protocol-worker-2/src/handlers/block_handler/retrieval.rs

```diff
@@ -0,0 +1,1280 @@
+    fn on_block_header_received(
+        &mut self,
+        from_peer_id: PeerId,
+        block_id: BlockId,
+        header: SecuredHeader,
+    ) -> Result<(), ProtocolError> {
+        if let Some(info) = self.block_wishlist.get(&block_id) {
+            if info.header.is_some() {
+                warn!(
+                    "Peer {} sent us header for block id {} but we already received it.",
+                    from_peer_id, block_id
+                );
+                if let Some(asked_blocks) = self.asked_blocks.get_mut(&from_peer_id) {
+                    if asked_blocks.contains_key(&block_id) {
+                        asked_blocks.remove(&block_id);
+                        {
+                            let mut cache_write = self.cache.write();
+                            cache_write.insert_blocks_known(
+                                &from_peer_id,
+                                &[block_id],
+                                false,
+                                Instant::now(),
+                            );
+                        }
+                    }
+                }
+
+                return Ok(());
+            }
+        }
+        if let Err(err) = self.note_header_from_peer(&header, &from_peer_id) {
+            warn!(
+                "peer {} sent us critically incorrect header through protocol, \
+                which may be an attack attempt by the remote node \
+                or a loss of sync between us and the remote node. Err = {}",
+                from_peer_id, err
+            );
+            let _ = self.ban_node(&from_peer_id);
+            return Ok(());
+        };
+        if let Some(info) = self.block_wishlist.get_mut(&block_id) {
+            info.header = Some(header);
+        }
+
+        // Update ask block
+        // Maybe this code is useless as it's been done just above but in a condition that should cover all cases where it's useful
+        // to do this. But maybe it's still trigger there it need verifications.
+        let mut set = PreHashSet::<BlockId>::with_capacity(1);
+        set.insert(block_id);
+        self.remove_asked_blocks_of_node(&set)?;
+        Ok(())
+    }
+
+    /// Perform checks on a header,
+    /// and if valid update the node's view of the world.
+    ///
+    /// Returns a boolean representing whether the header is new.
+    ///
+    /// Does not ban the source node if the header is invalid.
+    ///
+    /// Checks performed on Header:
+    /// - Not genesis.
+    /// - Can compute a `BlockId`.
+    /// - Valid signature.
+    /// - Absence of duplicate endorsements.
+    ///
+    /// Checks performed on endorsements:
+    /// - Unique indices.
+    /// - Slot matches that of the block.
+    /// - Block matches that of the block.
+    pub(crate) fn note_header_from_peer(
+        &mut self,
+        header: &SecuredHeader,
+        from_peer_id: &PeerId,
+    ) -> Result<Option<(BlockId, bool)>, ProtocolError> {
```

Comment:
a Result that can not return an error :disappointed_relieved:

Reply:
I don't want to change this for now before we can verify with the tests. Added a todo

Reply:
ok